core/trino-main/src/main/java/io/trino/server/ServerInfoResource.java

```diff
@@ -45,21 +50,30 @@
 @Path("/v1/info")
 public class ServerInfoResource
 {
+    private static Logger log = Logger.get(ServerInfoResource.class);
+
     private final NodeVersion version;
     private final String environment;
     private final boolean coordinator;
     private final GracefulShutdownHandler shutdownHandler;
     private final StartupStatus startupStatus;
+    private final DecommissionHandler decommissionHandler;
     private final long startTime = System.nanoTime();
+    private final AtomicBoolean startupComplete = new AtomicBoolean();
+    private NodeState currState = NodeState.ACTIVE;
```

Comment:
I don't really get the point here. ServerInfoResource has following APIs:
- public Response updateState(NodeState state)
- public NodeState getServerState()
It handles node state change request. There was no currState field before because there was only two states (ACTIVE or SHUTTING_DOWN) and the code indirectly tracks the current state through shutdownHandler.isShutdownRequested(), which is one direction (once shutdown, impossible to be back to active). With DECOMMISSIONING and DECOMMISSIONED, there are more states and no longer one directional, so the current state need to be tracked somewhere, to efficiently and correctly handle state change. If the current state is maintained inside decommissionHandler, then code inside updateState() would have to consult decommissionHandler for current state, which is not desired as I see code readability decreases as a result.

Reply:
I am merely stating the fact that the current design of this class does not match what we do in other parts of the code.
The common design principle is that `*Resource.java` class does not have any complex logic, nor they contain state. Those are used to expose information/capture requests to change information, which is passed deeper into specialized, smarter entities.
Also with current design you are duplicating some pieces of information making it non-obvious what is the source of truth (currState==SHUTTING_DOWN vs GracefulShutdownHandler.shutdownRequested)

With that in mind, we should not have `currState` here.
Also, we should probably not have the state machine logic.

It feels to me that we should have a new class that encapsulates `gracefulShutdown` and `decomissioning` handler.
Maybe those two should actually be merged into one? Maybe just `DecomissioningHandler` which also handles graceful shutdown requests?

Reply:
Got it. I will see what a combined class can handle it. The idea is clear, and likely will work out, despite details like proper class name TBD.

Reply:
Tried to combine GracefulShutdownHandler and DecommissionHandler into UpdateNodeStateHandler, which is used by ServerInfoResource.

Reply:
Sorry I forget to "git add UpdateNodeStateHandler.java" in my previous update

Reply:
Please let me know when it is ready for next round of review (is it already? CI is pretty red)

Reply:
@losipiuk Maybe it's ready for next round of review:
1. The current CI / check-commits error appears related to rebase during CI. I am afraid of messing up this PR if I try to rebase now, so will defer doing it after next round.
2. Exposing DN and DD in web UI is blocked as I am not sure how to deal with React and JS (see my other comments). These two counters could be added separately or later (once I know how to edit dist/index.js).